### PR TITLE
added allowprereleasedependencies flag in pack

### LINKED ISSUE #3938: Throw error when trying to create a non prerelease package with prerelease dependencies
Moved from https://github.com/dotnet/cli/issues/3844 on behalf of @niemyjski.

----

You can have a package (1.0.0) that depends on any prerelease package and it will create the 
package but you are unable to install it unless you specify a -pre flag.... This is a bug that just bit us.

### PR DESCRIPTION
**Pull request by rohit21agrawal**
Fixes : https://github.com/NuGet/Home/issues/3938

This fixes the issue in which NuGet pack (and now dotnet pack too) led to error if a stable package had prerelease dependencies.

This PR moves that logic to Package Analysis rules and therefore is downgraded as a warning. Package analysis can be suppressed by the -NoPackageAnalysis command line flag (or equivalent msbuild property).

Sample output :

D:\DevTools\dotnet-dev-win-x64.latest\sdk\1.0.0-preview5-004275\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(75,5): warning : Issue found with package 'packtest'. [D:\temp\packtest\packtest.csproj]
D:\DevTools\dotnet-dev-win-x64.latest\sdk\1.0.0-preview5-004275\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(75,5): warning : Issue: Prerelease dependency in stable package. [D:\temp\packtest\packtest.csproj]
D:\DevTools\dotnet-dev-win-x64.latest\sdk\1.0.0-preview5-004275\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(75,5): warning : Description: A stable release of a package should not have a prerelease dependency. [D:\temp\packtest\packtest.csproj]
D:\DevTools\dotnet-dev-win-x64.latest\sdk\1.0.0-preview5-004275\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(75,5): warning : Solution: Either modify the version spec of dependency "NuGet.Build.Tasks.Pack [4.0.0-zlocal-*, )" or update the version field in the nuspec. [D:\temp\packtest\packtest.csproj]

CC: @emgarten @alpaix @jainaashish @joelverhagen @mishra14 @rrelyea @yishaigalatzer

### CODE REVIEW

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/Pack.targets [hunk 1]
```diff
@@ -19,6 +19,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   <UsingTask TaskName="NuGet.Build.Tasks.Pack.PackNuspecTask" AssemblyFile="$(NugetTaskAssemblyFile)" />
 
   <PropertyGroup>
+    <AllowPrereleaseDependencies Condition=" '$(AllowPrereleaseDependencies)' == '' ">true</AllowPrereleaseDependencies>
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/Pack.targets]: I would expect this to be false by default

File: src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs [hunk 2]
```diff
@@ -31,6 +31,7 @@ public class PackArgs
         public bool Symbols { get; set; }
         public bool Tool { get; set; }
         public string Version { get; set; }
+        public bool AllowPrereleaseDependenciesInProductionPackage { get; set; }
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Commands/CommandArgs/PackArgs.cs]: how about ``AllowPrereleaseDependenciesInStablePackage``?

I'm not sure Production is the opposite of pre-release.

**Review by emgarten — CHANGES_REQUESTED** [review]
waiting on default

**Comment by rohit21agrawal** [comment]
@emgarten as per our discussion, i have changed this to a package analysis rule.

File: src/NuGet.Core/NuGet.Commands/Rules/AnalysisResources.resx [hunk 3]
```diff
@@ -153,6 +153,15 @@
   <data name="InvalidFrameworkTitle" xml:space="preserve">
     <value>Invalid framework folder.</value>
   </data>
+  <data name="InvalidPrereleaseDependency_Description" xml:space="preserve">
+    <value>A stable release of a package should not have a prerelease dependency.</value>
+  </data>
+  <data name="InvalidPrereleaseDependency_Solution" xml:space="preserve">
+    <value>Either modify the version spec of dependency "{0}" or update the version field in the nuspec.</value>
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Commands/Rules/AnalysisResources.resx:160]: nit: typically single quotes are used around fields like this

**Review by emgarten — APPROVED** [review]

**Comment by Henrik-R** [comment]
When will this fix be released?

I couldn't find it in any of the 4.0.0 release notes and it isn't attached to a milestone.

Have tested with 4.0.0 RTM and it seems to have the described behaviour

**Comment by shytikov** [comment]
Ok, guys, haven't you considered that this function (to fail pack project with pre-release dependencies) is actually used by someone in production?

Is it possible to create switch to keep old behavior untouched?

**Pull request merged**